dart/flat_buffers.dart

```diff
@@ -0,0 +1,1095 @@
+
+  @override
+  List<bool> read(BufferContext bc, int offset) =>
+      new _FbBoolList(bc, bc.derefObject(offset));
+}
+
+/**
+ * The reader of booleans.
+ */
+class BoolReader extends Reader<bool> {
+  const BoolReader() : super();
+
+  @override
+  int get size => 1;
+
+  @override
+  bool read(BufferContext bc, int offset) => bc._getInt8(offset) != 0;
+}
+
+/**
+ * Buffer with data and some context about it.
+ */
+class BufferContext {
+  final ByteData _buffer;
+
+  factory BufferContext.fromBytes(List<int> byteList) {
+    Uint8List uint8List = _asUint8List(byteList);
+    ByteData buf = new ByteData.view(uint8List.buffer, uint8List.offsetInBytes);
+    return new BufferContext._(buf);
+  }
+
+  BufferContext._(this._buffer);
+
+  int derefObject(int offset) {
+    return offset + _getUint32(offset);
+  }
+
+  Uint8List _asUint8LIst(int offset, int length) =>
+      _buffer.buffer.asUint8List(_buffer.offsetInBytes + offset, length);
+
+  double _getFloat64(int offset) =>
+      _buffer.getFloat64(offset, Endianness.LITTLE_ENDIAN);
+
+  double _getFloat32(int offset) =>
+      _buffer.getFloat32(offset, Endianness.LITTLE_ENDIAN);
+
+  int _getInt32(int offset) =>
+      _buffer.getInt32(offset, Endianness.LITTLE_ENDIAN);
+
+  int _getInt16(int offset) =>
+      _buffer.getInt16(offset, Endianness.LITTLE_ENDIAN);
+
+  int _getInt8(int offset) => _buffer.getInt8(offset);
+
+  int _getUint32(int offset) =>
+      _buffer.getUint32(offset, Endianness.LITTLE_ENDIAN);
+
+  int _getUint16(int offset) =>
+      _buffer.getUint16(offset, Endianness.LITTLE_ENDIAN);
+
+  int _getUint8(int offset) => _buffer.getUint8(offset);
+
+  /// If the [byteList] is already a [Uint8List] return it.
+  /// Otherwise return a [Uint8List] copy of the [byteList].
+  static Uint8List _asUint8List(List<int> byteList) {
+    if (byteList is Uint8List) {
+      return byteList;
+    } else {
+      return new Uint8List.fromList(byteList);
+    }
+  }
+}
+
+/**
+ * Class that helps building flat buffers.
+ */
+class Builder {
+  final int initialSize;
+
+  /// The list of existing VTable(s).
+  final List<_VTable> _vTables = <_VTable>[];
+
+  ByteData _buf;
+
+  /// The maximum alignment that has been seen so far.  If [_buf] has to be
+  /// reallocated in the future (to insert room at its start for more bytes) the
+  /// reallocation will need to be a multiple of this many bytes.
+  int _maxAlign;
+
+  /// The number of bytes that have been written to the buffer so far.  The
+  /// most recently written byte is this many bytes from the end of [_buf].
+  int _tail;
+
+  /// The location of the end of the current table, measured in bytes from the
+  /// end of [_buf], or `null` if a table is not currently being built.
+  int _currentTableEndTail;
+
+  _VTable _currentVTable;
+
+  /// Map containing all strings that have been written so far.  This allows us
+  /// to avoid duplicating strings.
+  Map<String, Offset<String>> _strings = <String, Offset<String>>{};
```

Comment:
this looks expensive. In C++, string pooling is optional, is that the case here?

Reply:
it is not optional here.  This map is used by `writeString` to achieve string interning in resultant file (and it is only used in write operations obviously).  I'm not sure it's much more expensive than writing duplicates of the string if a string is duplicated. It would not be difficult to allow consumers of `writeString` to opt out of this (or opt in), but I'd think it'd make more sense to opt out.

Reply:
Wether duplicates are likely really depends on the use case. There are plenty of cases where strings don't duplicate at all, and then you end up copying all strings into a map for no reason. FlatBuffers should by default be as low allocation as possible. Since in C++ it is opt-in for that reason, I suggest we do the same in Dart (most other languages have no string pooling feature at this time).

Reply:
I don't have opinion on whether making canonicalization opt-in or opt-out, we generate code that interacts with flat buffers, so we can change it easily.

We store information about Dart packages in flat buffers, and names of classes and methods often duplicates across files in the same package, so we decided to implement this canonicalization.

Reply:
I've made it opt-in to match C++ API